libdrgn/python/program.c

```diff
@@ -504,6 +504,13 @@ py_symbol_find_fn(const char *name, uint64_t addr,
 		  enum drgn_find_symbol_flags flags, void *arg,
 		  struct drgn_symbol_result_builder *builder)
 {
+	// Fast path for SymbolIndex: don't bother converting to and from Python
+	// types, as this is a C finder.
+	if (PyObject_TypeCheck(PyTuple_GET_ITEM(arg, 1), &SymbolIndex_type)) {
```

Comment:
Fundamentally we don't actually care about subclasses of `SymbolIndex_type`, because it actually can't be subclassed right now:

```python
>>> from drgn import SymbolIndex
>>> class MySymbolIndex(SymbolIndex):
...     def __call__(self, *args, **kwargs):
...         print("I'm a problem for you!")
...         super().__call__(self, *args, **kwargs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: type '_drgn.SymbolIndex' is not an acceptable base type
```

And even if it could be inherited from, that would actually be a problem, since we would ignore any overridden `__call__()` method with this fast path.

Would it be any safer to do `Py_TYPE(PyTuple_GET_ITEM(arg, 1)) == &SymbolIndex_type`? This would avoid the logic for checking if it's a subclass altogether.

Reply:
Yeah, I think we can get away with that.